proto/devices.proto

```diff
@@ -94,12 +94,28 @@ message Device {
 
         // The condition matches when the field matches all conditions in the given query.
         // Fields in the sub-query are relative to value resolved by the parent condition.
+        // Sub-query conditions with empty field properties are compared against all nested values
+        // which is like using the value matched by this conditions field as the root for the sub-query.
         // For non-repeated fields, this is equivalent to providing multiple conditions on the root query.
-        // For repeated fields, or sub fields of those, there must exist at least one entry where all sub-query conditions match.
+        // For repeated fields, or sub fields of those, items match if all sub-query conditions match.
         Query matches = 50;
+      }
 
-        // there's room here for additional rhs values for the field, but we don't need any
+      // Matcher specifies how the overall condition behaves when matched against a field.
+      enum Matcher {
+        // When unspecified, the matching behaviour is equivalent to using ANY.
+        MATCHER_UNSPECIFIED = 0;
+        // The condition matches if any value matches the condition.
+        // This is the default behaviour when unspecified.
+        // For repeated values, any item matching the condition is sufficient for the record to match.
+        ANY = 1;
+        // The condition matches only if all value match the condition.
+        // For non-repeated values this is equivalent to ANY.
+        // For repeated values, all items must match the condition for the record to match.
+        ALL = 2;
       }
+      // Matcher controls how values are matched against this condition.
+      Matcher matcher = 100;
```

Comment:
This extra property on the Query isn't very consistent with the rest of the Query design. All other variations of matching are achieved with seperate cases in the `value` oneof. Couldn't we have `Query all_match` and `Query any_match` here to achieve the functionality?

I also don't think the docs make clear that the Matcher is only relevant for repeated fields.

Reply:
I was hoping to avoid that distinction here as if we ever add a `NOT` concept it would live here as a matcher. The matcher technically applies to both singular and repeated values, it's just in the former case the observed behaviour will be identical for both ANY and ALL.